### PR TITLE
Stub implementation of SVG basic geometric shapes

### PR DESCRIPTION
**Pull request by marinho**
This PR is aimed to solve https://github.com/jsdom/jsdom/issues/2128, mostly as a stub implemention, as it doesn't include full implementation for the given classes.

Instead, it provides the SVG geometric classes and their dependencies, including all attributes and methods as defined in specs, even though such methods have empty code.

This is strictly following the [official specs](https://www.w3.org/TR/SVG/), but as it doesn't provide much information about most of those classes, I also follow [Mozilla WebIDL files](https://dxr.mozilla.org/mozilla-beta/source/dom/webidl/), excluding Mozilla specific attributes.

I didn't enable **DIR: svg** entry in `tu-run.yaml`, as they don't pass, raising an error for missing **WebIDL2**, and they are unlike to pass anyway as my methods are empty.

I splitted the work into 2 commits, as I'm not sure if the attributes implementation based on Mozilla WebIDL is desirable. Please let me know if I should remove them.

It can be seen as a continuation of https://github.com/jsdom/jsdom/pull/2011

### CODE REVIEW

**Comment by marinho** [comment]
It seems that errors happening on `referrer.any.html` and `security/001.html` aren't making Travis to fail, right?

So I guess it's caused by linting errors at the beginning of the output?

I'm going to take care of them.

**Comment by TimothyGu** [comment]
Also just in general, we always want to follow the editor's draft of specifications, since that is what browser developers actually look at when implementing features. In other words, URLs that look like `https://www.w3.org/TR/` are pretty much never appropriate. `https://dev.w3.org/` and `https://w3c.github.io` however, are generally fine, except for [HTML](https://annevankesteren.nl/2016/01/film-at-11) and [some other specs maintained by WHATWG](https://wiki.whatwg.org/wiki/Fork_tracking).

**Comment by marinho** [comment]
thanks for the review. I'm going to make corrections and push it again soon!

**Comment by marinho** [comment]
@TimothyGu one question on empty interfaces missing implementation, what's the general guidance then?

Does it make sense to keep tagged element interfaces (i.e. `SVGPathElement`, `SVGRectElement`, etc.) with their attributes and comment their methods out?

Regarding to interfaces which aren't tagged elements (i.e. `SVGAnimatedLength`), should I replace them by some wildcard type, comment them out or implement them properly?

thanks

**Comment by TimothyGu** [comment]
> Does it make sense to keep tagged element interfaces (i.e. `SVGPathElement`, `SVGRectElement`, etc.) with their attributes and comment their methods out?

If you don't want to implement the methods, then yes comment them out in the IDL files.

If the attributes work, then sure, keeping them is fine. Otherwise, comment them out too.

> Regarding to interfaces which aren't tagged elements (i.e. `SVGAnimatedLength`), should I replace them by some wildcard type, comment them out or implement them properly?

If you have the time and ability to implement them, then implement them properly.
 (`SVGAnimatedLength` is one such class where the things needed to be implemented are probably too plentiful to worth the effort.) If not, then remove them, and comment out the attributes in the IDL for Element classes that use them.

**Comment by domenic** [comment]
In general I would prefer we not include stub implementations at all. Unless we implement a significant portion of a class, it's better to just leave it as SVGElement, as that indicates more clearly to consumers that it's not supported.

It would help if there was a prose explanation of this change in terms of what features it implements.

**Comment by marinho** [comment]
@domenic, I was thinking that too after I thought on @TimothyGu comments.

My work plan is to get rid of `SVGAnimatedPathData`, `SVGPathSeg`, `SVGPathSegList` and related attributes and implement the rest as much as possible. I'm not sure what challenges I will face though (still couldn't work on it). I will probably have an update at the end of next week.

**Comment by marinho** [comment]
@domenic and @TimothyGu 

I started some implementation, but I'd like to have your feedback before I go ahead. Please take a look at the latest 3 commits I made (I didn't touch on the 3 commits that were there before), considering the following concerns:

- I didn't update it to latest jsdom master yet
- I added 2 new dependencies (d3-polygon and svg-pathdata
- wrote unit tests (tuwpt) instead of enabling tests based of wpt (for the moment)
- Although path data is no longer exposed by Chrome SVG API, the SVG 2 draft state that all SVG shapes still resolve into paths, so that `getPathData` is present only for internal matter and I plan to hide it as a private method
- The SVG present in wpt tests is not easy to cover, as it's more focused on transitions, animations and other features. I'm going to investigate how to enable some of them

If you give a positive feedback, I'm going to implement the basic methods according to SVG2 draft for these shapes, but I would like to know if it's better push it as smaller steps or wait until I have all the shapes ready.

PS: By "shapes", I mean these elements:

- `SVGCircleElement`
- `SVGEllipsisElement`
- `SVGLineElement`
- `SVGPolygonElement`
- `SVGPolylineElement`
- `SVGPathElement`
- `SVGRectElement`

best,

File: lib/jsdom/living/register-elements.js [hunk 1]
```diff
@@ -356,6 +356,10 @@ const mappings = {
       file: require("./generated/SVGGeometryElement.js"),
       tags: []
     },
+    SVGLength: {
```
Comment (domenic) [inline lib/jsdom/living/register-elements.js]: This isn't an element so doesn't belong here.

File: test/web-platform-tests/to-upstream/svg/svgrectelement.html [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  assert_true(rect instanceof SVGElement);
+  assert_true(rect instanceof SVGGraphicsElement);
+  assert_true(rect instanceof SVGGeometryElement);
+  assert_true(rect instanceof SVGRectElement);
+  assert_false(rect instanceof HTMLElement);
+}, "Basic properties");
+
+test(() => {
+  const rect = document.querySelector("#aRect");
+  const pathData = rect.getPathData();
```
Comment (domenic) [inline test/web-platform-tests/to-upstream/svg/svgrectelement.html]: This isn't in any spec so shouldn't be tested in WPTs.

File: test/web-platform-tests/to-upstream/svg/svgpathelement.html [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+  assert_true(path instanceof SVGElement);
+  assert_true(path instanceof SVGGraphicsElement);
+  assert_true(path instanceof SVGGeometryElement);
+  assert_true(path instanceof SVGPathElement);
+  assert_false(path instanceof HTMLElement);
+}, "Basic properties");
+
+test(() => {
+  const path = document.querySelector("#aPath");
+  const pathData = path.getPathData();
```
Comment (domenic) [inline test/web-platform-tests/to-upstream/svg/svgpathelement.html]: This isn't in any spec so shouldn't be tested in WPTs

**Comment by seblucas** [comment]
Don't know if somebody is still working of this, but I'm also using latest d3.js and I'm also needing this ;)

**Comment by domenic** [comment]
Closing this as it seems abandoned :(.

**Comment by ulfaslak** [comment]
Such a shame this never went anywhere. @marinho's original problem was that when JSDOM does not have support for these geometric shapes, there is no clean way to write tests (with Jest) for code that uses them. Many people have this problem. In my case I (like @marinho) need a polyfill, so the fix is [a monkeypatch](https://github.com/facebook/jest/issues/5379#issuecomment-360044161).

Also, just because it struck me as a bystander: the attitude (especially @domenic's) towards OP in these interactions was hostile. Nobody wants to put time into a space like that.

**Pull request closed without merging**